src/snowflake/snowpark/dataframe.py

```diff
@@ -1390,20 +1391,24 @@ def filter(self, expr: ColumnOrSqlExpr) -> "DataFrame":
 
         Args:
             expr: a :class:`Column` expression or SQL text.
+            _ast_stmt: when invoked internally, supplies the AST to use for the resulting dataframe.
 
         :meth:`where` is an alias of :meth:`filter`.
         """
         # AST.
-        stmt = self._session._ast_batch.assign()
-        ast = stmt.expr.sp_dataframe_filter
-        self.set_ast_ref(ast.df)
-        set_src_position(ast.src)
-        if isinstance(expr, Column):
-            pass  # TODO
-        elif isinstance(expr, str):
-            ast.condition.sp_column_sql_expr.sql = expr
+        if _ast_stmt is None:
+            stmt = self._session._ast_batch.assign()
+            ast = stmt.expr.sp_dataframe_filter
```

Comment:
should we use here the `with_src_position` helper as well?

Reply:
Yes. However, I'd say this is outside of the scope of this diff (i.e. we need to update that in quite a few places). This code had to be made conditional to support nested invocations.